integrations/upgrade/js-config.test.ts

```diff
@@ -468,3 +642,218 @@ test(
+         *
+         * It is recommended to migrate your project, and delete this layer once you
+         * are done.
+         *
+         * - All elements with a 'border' class will default to 'currentColor' in
+         *   Tailwind CSS v4, instead of 'border-gray-200' in Tailwind CSS v3.
+         * - All inputs, selects, and textareas will have a 'border-width: 1px' in
+         *   Tailwind CSS v4, instead of 'border-width: 0' in Tailwind CSS v3.
+         */
+        @layer base {
+          *,
+          ::after,
+          ::before,
+          ::backdrop,
+          ::file-selector-button {
+            border-color: var(--border-color, var(--color-gray-200, currentColor));
+          }
+
+          input:where(:not([type='button'], [type='reset'], [type='submit'])),
+          select,
+          textarea {
+            border-width: 0;
+          }
+        }
+
+        @theme {
+          --border-color: oklch(0.623 0.214 259.815);
+        }
+        "
+      `)
+    },
+  )
+
+  test(
+    'migrate border compatibility if a custom border color is used, that matches the default v4 border color',
+    {
+      fs: {
+        'package.json': json`
+          {
+            "dependencies": {
+              "@tailwindcss/upgrade": "workspace:^"
+            }
+          }
+        `,
+        'tailwind.config.ts': ts`
+          import { type Config } from 'tailwindcss'
+
+          export default {
+            theme: {
+              extend: {
+                borderColor: ({ colors }) => ({
+                  DEFAULT: 'currentColor',
+                }),
+              },
+            },
+          } satisfies Config
+        `,
+        'src/input.css': css`
+          @tailwind base;
+          @tailwind components;
+          @tailwind utilities;
+        `,
+      },
+    },
+    async ({ exec, fs }) => {
+      await exec('npx @tailwindcss/upgrade')
+
+      expect(await fs.dumpFiles('src/**/*.css')).toMatchInlineSnapshot(`
+        "
+        --- src/input.css ---
+        @import 'tailwindcss';
+
+        /**
+         * The default border styles in Tailwind CSS v4 changed compared to 
+         * Tailwind CSS v3. To make sure your project looks the same, we've added
+         * these compatibility styles to your project.
+         *
+         * It is recommended to migrate your project, and delete this layer once you
+         * are done.
+         *
+         * - All elements with a 'border' class will default to 'currentColor' in
+         *   Tailwind CSS v4, instead of 'border-gray-200' in Tailwind CSS v3.
+         * - All inputs, selects, and textareas will have a 'border-width: 1px' in
+         *   Tailwind CSS v4, instead of 'border-width: 0' in Tailwind CSS v3.
+         */
+        @layer base {
+          input:where(:not([type='button'], [type='reset'], [type='submit'])),
+          select,
+          textarea {
+            border-width: 0;
+          }
+        }
```

Comment:
Notice that the default styles for the `border-color` are not here because they resolve to `currentColor` which is also the default in v4. Therefore if we would've added it, it would be additional CSS doing the same thing.